src/cdr/resolver.go

```
@@ -0,0 +1,271 @@
+		if err != nil {
+			return nil, err
+		}
+		r.cache.put(ref, data)
+		return io.NopCloser(bytes.NewReader(data)), nil
+	}
+	var rc io.ReadCloser
+	var err error
+	switch body := ref.Body.(type) {
+	case *Ref_Http:
+		rc, err = r.derefHTTP(ctx, body.Http)
+		if err != nil {
+			return nil, err
+		}
+	case *Ref_ContentHash:
+		rc, err = r.derefContentHash(ctx, body.ContentHash, false)
+		if cache && IsImmutable(ref) {
+			rc, err = cacheRC(rc)
+		}
+	case *Ref_Cipher:
+		rc, err = r.derefCipher(ctx, body.Cipher, false)
+		if err != nil {
+			return nil, err
+		}
+		if cache && IsImmutable(ref) {
+			rc, err = cacheRC(rc)
+		}
+	case *Ref_Compress:
+		rc, err = r.derefCompress(ctx, body.Compress, false)
+		if err != nil {
+			return nil, err
+		}
+		if cache && IsImmutable(ref) {
+			rc, err = cacheRC(rc)
+		}
+	case *Ref_Slice:
+		rc, err = r.derefSlice(ctx, body.Slice, cache)
+	case *Ref_Concat:
+		rc, err = r.derefConcat(ctx, body.Concat, cache)
+	default:
+		return nil, errors.Errorf("unsupported Ref variant %v: %T", ref.Body, ref.Body)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return rc, nil
+}
+
+func (r *Resolver) derefHTTP(ctx context.Context, ref *HTTP) (io.ReadCloser, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, ref.Url, nil)
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range ref.Headers {
+		req.Header.Add(k, v)
+	}
+	// TODO: Check response code.
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+func (r *Resolver) derefContentHash(ctx context.Context, ref *ContentHash, cache bool) (io.ReadCloser, error) {
+	innerRc, err := r.deref(ctx, ref.Inner, cache)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: add a maximum limit here
+	data, err := io.ReadAll(innerRc)
+	if err != nil {
+		return nil, err
+	}
+	var h hash.Hash
+	switch ref.Algo {
+	case HashAlgo_BLAKE2b_256:
+		h, err = blake2b.New256(nil)
+		if err != nil {
+			panic(err)
```

Are we sure we want to panic here instead of just returning an error up the stack? Won't this crash the entire pod?